forge-game/src/main/java/forge/game/cost/CostAdjustment.java

```diff
@@ -208,6 +209,22 @@ else if (stAb.getParam("Mode").equals("SetCost")) {
         // need to reduce generic extra because of 2 hybrid mana
         cost.decreaseGenericMana(sumGeneric);
 
+        if (sa.isSpell() && sa.isOptionalCostPaid(OptionalCost.ReduceW)) {
```

Comment:
The current code doesn't check for which ability does you pay the cost

Meaning if you get 2 defiler of the same color, and you pay life for one, the other would still reduce the mana shard

Reply:
OK, I understand – in addition to that problem, there is currently another problem: if you pay the additional cost for 2 defilers of the same color, it only reduces by one pip.

Reply:
I couldn't figure out the best way to follow your suggestion... perhaps you or @tool4ever can advise (or revise once this PR is done)

For now just logging the pips on the sa does solve the problem cases we found

Reply:
OptionalCostValue currently stores Type and Cost, it need to also store staticAbilityId with default either null or -1

isOptionalCostPaid could be extended to check for the staticAbilityId too

then when OptionalCost static is checked in CostAdjustment, it could check for if this specific ability was used

Reply:
I was able to extend OptionalCostValue to include the staticAbilityID but beyond that I am lost on this implementation. Not sure if this PR needs to be closed or if it's close enough to be a good start for someone else to fix when they have time.

Reply:
Marked this for follow-up in an Issue